Homework 2

nikita-talalai opened the pull request.

nikita-talalai committed:
Add solution for hw2.task1

nikita-talalai committed:
Delete solution for hw2.task1

nikita-talalai committed:
Add and move solution for hw2.task1 to folder hw2

nikita-talalai committed:
Add tests for hw2.task1

nikita-talalai committed:
Add solution for task4

DedSec256 commented on hw2/task1/Tests.fs:
```diff
@@ -0,0 +1,25 @@
+﻿module hw2.Tests
```
Не хватает проверки пустого списка и списка с отрицательными числами.

nikita-talalai committed:
Add tests for task1

nikita-talalai committed:
Add solution and tests for task2 (Map for tree)

Map realization for tree in CPS

nikita-talalai committed:
Add solution and tests for task3 (Parse tree)

nikita-talalai committed:
Clean file

DedSec256 commented on hw2/task2/Program.fs:
```diff
@@ -0,0 +1,13 @@
+let mapTree mapping tree =
+    let rec innerMap tree cont =
+        match tree with
+        | Node(x, l, r) ->
+            innerMap l (fun left -> innerMap r (fun right -> cont (Node(mapping x, left, right))))
+        | Empty -> cont Empty
```
За старания 5ка, но в реальной жизни более простое решение "в лоб" без страшной хвостовой рекурсии с CPS здесь подошло бы лучше, вероятно.

DedSec256 commented on hw2/task3/Program.fs:
```diff
@@ -0,0 +1,23 @@
+
+type Expression =
+    | Const of int
+    | Plus of Expression * Expression
+    | Minus of Expression * Expression
+    | Times of Expression * Expression
+    | Divide of Expression * Expression
```
Можно упростить до BinaryExpression

DedSec256 commented on hw2/task4/Program.fs:
```diff
@@ -0,0 +1,11 @@
+    | n when n <= 1 -> false
+    | n when n = 2 -> true
+    | n ->
+        2 :: [3 .. 2 .. (n |> float |> sqrt |> int) + 1]
+        |> List.filter (fun x -> n % x = 0)
+        |> List.isEmpty 
```
лучше не аллоцировать целые списки для такой простой задачи

DedSec256 commented on hw2/task4/Tests.fs:
```diff
@@ -0,0 +1,22 @@
+        
+[<Test>]
+let ``Expected generation`` () =
+    primeGenerator() |> Seq.skip 10 |> Seq.take 10 |> should equalSeq <| seq {31; 37; 41; 43; 47; 53; 59; 61; 67; 71}
```
самый полезный тест, остальные сравнивают функцию isPrime саму с собой, что некорректно.

nikita-talalai committed:
test: add test projects

Create test project for each task. Move files with tests to corresponding test project.

nikita-talalai committed:
test: remove redundant test cases

nikita-talalai committed:
fix: remove potential risk out of memory exception
